Configure EC2 instance metadata options

muraee opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->
<!-- If this PR is still work-in-progress and is being open for visibility please prefix the title with `WIP:` -->

**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind regression
/kind support
-->

**What this PR does / why we need it**:
Allows configuration of EC2 instance metadata options to add IMDSv2 support. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3744

**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [ ] squashed commits
- [x] includes documentation
- [ ] adds unit tests
- [x] adds or updates e2e tests

**Release Note**:
```
Add support to allow users to configure EC2 instance metadata options
```

Skarlso commented:
/ok-to-test

muraee commented:
@Skarlso do I need to include the new API field in `vbeta1` as well?

Skarlso commented:
>@Skarlso do I need to include the new API field in vbeta1 as well?

Nope, there is like a bunch of stuff regarding conversionts.

I'll include something tomorrow, I'm exhausted for today. Try looking up the git history of the conversion.go file and *_conversion.go files in the v1 folder. That will give you some hints as to how to proceed with them.

muraee commented:
> > @Skarlso do I need to include the new API field in vbeta1 as well?
> 
> Nope, there is a bunch of stuff regarding conversions.
> 
> I'll include something tomorrow, I'm exhausted for today. Try looking up the git history of the conversion.go file and *_conversion.go files in the v1 folder. That will give you some hints as to how to proceed with them.

I fixed the conversion issues, I hope this is the correct way.

Skarlso commented:
Nope, sadly, it's not. :D I'll try to point you to the right thing today. Sorry, I'm just so frigging busy.

muraee commented:
@Skarlso could you please take a look if you have got time this week. Thanks :)

cnmcavoy commented on api/v1beta1/conversion.go:
```diff
@@ -43,6 +43,14 @@ func Convert_v1beta2_NetworkStatus_To_v1beta1_NetworkStatus(in *v1beta2.NetworkS
 	return autoConvert_v1beta2_NetworkStatus_To_v1beta1_NetworkStatus(in, out, s)
 }
 
+func Convert_v1beta2_AWSMachineSpec_To_v1beta1_AWSMachineSpec(in *v1beta2.AWSMachineSpec, out *AWSMachineSpec, s conversion.Scope) error {
```
So what I believe is missing with the conversion that @Skarlso mentioned is that the conversion currently discards the new fields when converting down to v1beta1, so a conversion down to v1beta1 and back to v1beta2 will result in data loss.

To fix this, the additional fields that are not present on v1beta1 but exist on v1beta2 are encoded as a value on an annotation, and extracted back out when converting from v1beta1 to v1beta2. 

Here's an example from one of my earlier PR's where @Skarlso helped me previously with the same sort of conversion logic: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3757/files#r989941779 and `"sigs.k8s.io/controller-runtime/pkg/conversion"` contains a number of helper methods related to conversion.

Skarlso replied:
Thank you Cameron. Sorry, I have my hands full with trying to fix this frigging test infra. :D

muraee replied:
thanks for the hint @cnmcavoy

muraee commented:
@Skarlso all tests passed, could I have a review?

Skarlso commented:
Okay this is looking better now. I'll give it a better look later today.

Skarlso commented:
/assign

Ankitasw reviewed (commented):
Overall LGTM. Is there a way we could verify this in one of the tests in E2E?

Ankitasw commented:
@muraee could you please squash your commits ?

muraee commented:
@Ankitasw thanks for the review, I squashed the commits and added new docs.

This could be easily validated in e2e, by describing any instance created and checking the returned metadata options. I am not sure though where to add this check, as I am not familiar with the e2e framework used here.

muraee commented:
/retest-required

Ankitasw commented:
@muraee suppose [this](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/suites/unmanaged/unmanaged_functional_test.go#L160) is one of the tests. At the end of test, we are verifying few things, could you check if you could add the checks to describe metadata and see if some of the fields of metadata matches the expected.

muraee commented:
@Ankitasw please take a look if the test makes sense to you?
could you also please take a look on why the `pull-cluster-api-provider-aws-test ` test is failing, It seems like an error unrelated to these changes,

Ankitasw commented:
> could you also please take a look on why the pull-cluster-api-provider-aws-test  test is failing, It seems like an error unrelated to these changes

Its not unrelated, actually we added new if block in `awsmachine_controller.go` so cyclomatic complexity is increased for that function.

muraee commented on controllers/awsmachine_controller.go:
```diff
@@ -596,6 +596,12 @@ func (r *AWSMachineReconciler) reconcileNormal(_ context.Context, machineScope *
 			return ctrl.Result{}, err
 		}
 		conditions.MarkTrue(machineScope.AWSMachine, infrav1.SecurityGroupsReadyCondition)
+
```
sure

Ankitasw commented:
/override pull-cluster-api-provider-aws-apidiff-main

Ankitasw commented:
@muraee there is error in new assertion added, could you please check the e2e failure?

muraee commented:
/test pull-cluster-api-provider-aws-e2e

muraee commented:
/test pull-cluster-api-provider-aws-e2e

muraee commented:
/test pull-cluster-api-provider-aws-e2e

muraee commented:
/test pull-cluster-api-provider-aws-e2e

muraee commented:
@Ankitasw e2e passed when options specified explicitly. Seems the `Set_Defaults()` is not working, any ideas why?

Ankitasw commented:
@muraee could you please squash your commits before we do the merge?

Ankitasw commented:
@richardcase is this change ok to include now? or shall we wait for api-change?

wyike commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -961,6 +982,23 @@ func (s *Service) checkRootVolume(rootVolume *infrav1.Volume, imageID string) (*
+// ModifyInstanceMetadataOptions modifies the metadata options of the given EC2 instance.
+func (s *Service) ModifyInstanceMetadataOptions(instanceID string, options *infrav1.InstanceMetadataOptions) error {
+	input := &ec2.ModifyInstanceMetadataOptionsInput{
+		HttpEndpoint:            aws.String(string(options.HTTPEndpoint)),
```
When creating a cluster, I feel I meet a regression error with latest code related to the options here.  Paste the trace for your awareness...

```
E0308 07:31:06.070432      17 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 499 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x26b8c20, 0x440c890})
	/Users/yikew/Projects/pkg/mod/k8s.io/apimachinery@v0.25.5/pkg/util/runtime/runtime.go:75 +0xe0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:110 +0xf0
panic({0x26b8c20, 0x440c890})
	/usr/local/go/src/runtime/panic.go:890 +0x264
sigs.k8s.io/cluster-api-provider-aws/v2/pkg/cloud/services/ec2.(*Service).ModifyInstanceMetadataOptions(0x400134be60, {0x4001053fc8, 0x13}, 0x0)
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/pkg/cloud/services/ec2/instances.go:971 +0x34
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).ensureInstanceMetadataOptions(0x40005e06c0, {0x2ffb9b0, 0x400134be60}, 0x4000f734a0, 0x40011222c0)
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/controllers/awsmachine_controller.go:1121 +0xf0
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).reconcileOperationalState(0x40005e06c0, {0x2ffb9b0, 0x400134be60}, 0x40015731a0, 0x4000f734a0)
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/controllers/awsmachine_controller.go:616 +0x28c
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).reconcileNormal(0x40005e06c0, {0x40005b5090, 0x40010988e8}, 0x40015731a0, {0x2ffb7f0, 0x400020c310}, {0x3003088, 0x400020c310}, {0x3001be0, 0x400020c310}, ...)
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/controllers/awsmachine_controller.go:588 +0x1624
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).Reconcile(0x40005e06c0, {0x2fee338, 0x4000b39680}, {{{0x40016b7a20, 0x7}, {0x4001030840, 0x16}}})
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/controllers/awsmachine_controller.go:224 +0x93c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x4000330c80, {0x2fee338, 0x4000b39680}, {{{0x40016b7a20, 0x7}, {0x4001030840, 0x16}}})
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:121 +0xec
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x4000330c80, {0x2fee338, 0x4000b39680}, {0x27acba0, 0x40000f5f20})
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:320 +0x2f8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x4000330c80, {0x2fee290, 0x40001a4300})
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273 +0x2b8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234 +0xa8
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:230 +0x400
E0308 07:31:06.070538      17 controller.go:326] "Reconciler error" err="panic: runtime error: invalid memory address or nil pointer dereference [recovered]" controller="awsmachine" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSMachine" AWSMachine="default/c6-control-plane-bswb4" namespace="default" name="c6-control-plane-bswb4" reconcileID=18c05f9f-f837-4366-ab61-e6ce9c225db8
```

Ankitasw replied:
Weird that this is not caught in E2E tests. @muraee could you please take a look.

wyike replied:
It won't fail the cluster creation, only raise exceptions in the capa logs.

Ankitasw replied:
I think when instancemetadata options are not set by user, then it will throw the exceptions, I will raise another PR to make it optional

Ankitasw replied:
One more thing is we noticed another [issue](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4117) with awsmachine webhooks so that could be the reason, thius defaults are not getting set.
I will wait for that issue to fix first.

muraee replied:
thx @Ankitasw, yes it seems the defaults are not getting set for some reason, Although I tried that before and it was working.

cnmcavoy replied:
I believe I hit this exact issue in a separate change. Assuming this is the case, then using the defaulting webhook is not enough. This is the code I think is missing: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4112/files#diff-0b559bbd149f0e6d54d789235423f66fa8906fdc6ee9c99b9e85db912912011eR192

`SetDefaults_AWSMachineSpec` needs to be invoked from the AWSMachine controller reconciliation loop. 

The patch site is also very important, see the comment.
```
	// The defaulting must happen before `NewMachineScope` is called since otherwise we keep detecting
	// differences that result in patch operations.

	... do defaulting
```
Specifically, what is suspect is happening is existing AWSMachines that were created before your change can't receive the defaults, since they already exist. So we need to also set the defaults in the controller, at least until the resource is upgraded to another major apiVersion, then we can stop defaulting in the controller.

muraee replied:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4147 should fix it

wyike commented:
@muraee  I didn't see where bastion machine is configured to IMDSv2. From my testing, it is also IMDSv1. Could you help double confirm? Do we need it to be IMDSv2 as well?

muraee commented:
@wyike right, the bastion is not configured to use IMDSv2 in this change since it is created/managed by the awscluster_controller.
I am not sure if we want the bastion to default to IMDSv2. Please open an issue for further discussions.
